### PR TITLE
Empty/None string bugs fixed for cornell_movie DoubleTeacher.

### PR DESCRIPTION
**Pull request by ShaojieJiang**
**Patch description**
For a dialogue as follows:
```
x1 y1
x2 y2
x3
```
The `DoubleTeacher` will create two problematic training pairs. These problems is especially obvious when using `batch_size = 1`.
The first is `['', x1]`. With an empty source sequence, the program crashes due to `batch.text_vec` and `batch.label_vec` are both `None`.
The second is `[x3, None]`. When this happens, `eval_step` function is called even when `model.training == True`.

Please enter a clear and concise description of what your pull request does, and why
it is necessary. If your patch fixes an issue, please reference that issue here.

**Testing steps**
No

**Logs**
No

**Other information**
No

**Data tests (if applicable)**
```
running test
Searching for Unidecode==1.1.1
Best match: Unidecode 1.1.1
Processing Unidecode-1.1.1-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/Unidecode-1.1.1-py3.7.egg
Searching for tqdm==4.36.1
Best match: tqdm 4.36.1
Processing tqdm-4.36.1-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/tqdm-4.36.1-py3.7.egg
Searching for Sphinx==2.2.0
Best match: Sphinx 2.2.0
Processing Sphinx-2.2.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/Sphinx-2.2.0-py3.7.egg
Searching for scipy==1.3.1
Best match: scipy 1.3.1
Processing scipy-1.3.1-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/scipy-1.3.1-py3.7-macosx-10.7-x86_64.egg
Searching for scikit-learn==0.21.3
Best match: scikit-learn 0.21.3
Processing scikit_learn-0.21.3-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/scikit_learn-0.21.3-py3.7-macosx-10.7-x86_64.egg
Searching for requests-mock==1.7.0
Best match: requests-mock 1.7.0
Processing requests_mock-1.7.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/requests_mock-1.7.0-py3.7.egg
Searching for requests==2.22.0
Best match: requests 2.22.0
Processing requests-2.22.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/requests-2.22.0-py3.7.egg
Searching for regex==2019.8.19
Best match: regex 2019.8.19
Processing regex-2019.8.19-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/regex-2019.8.19-py3.7-macosx-10.7-x86_64.egg
Searching for pyzmq==18.1.0
Best match: pyzmq 18.1.0
Processing pyzmq-18.1.0-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/pyzmq-18.1.0-py3.7-macosx-10.7-x86_64.egg
Searching for pre-commit==1.18.3
Best match: pre-commit 1.18.3
Processing pre_commit-1.18.3-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/pre_commit-1.18.3-py3.7.egg
Searching for Pillow==6.2.0
Best match: Pillow 6.2.0
Processing Pillow-6.2.0-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/Pillow-6.2.0-py3.7-macosx-10.7-x86_64.egg
Searching for pexpect==4.7.0
Best match: pexpect 4.7.0
Processing pexpect-4.7.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/pexpect-4.7.0-py3.7.egg
Searching for numpy==1.17.2
Best match: numpy 1.17.2
Processing numpy-1.17.2-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/numpy-1.17.2-py3.7-macosx-10.7-x86_64.egg
Searching for nltk==3.4.5
Best match: nltk 3.4.5
Processing nltk-3.4.5-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/nltk-3.4.5-py3.7.egg
Searching for joblib==0.14.0
Best match: joblib 0.14.0
Processing joblib-0.14.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/joblib-0.14.0-py3.7.egg
Searching for h5py==2.10.0
Best match: h5py 2.10.0
Processing h5py-2.10.0-py3.7-macosx-10.7-x86_64.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/h5py-2.10.0-py3.7-macosx-10.7-x86_64.egg
Searching for GitPython==3.0.3
Best match: GitPython 3.0.3
Processing GitPython-3.0.3-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/GitPython-3.0.3-py3.7.egg
Searching for flake8-rst-docstrings==0.0.11
Best match: flake8-rst-docstrings 0.0.11
Processing flake8_rst_docstrings-0.0.11-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/flake8_rst_docstrings-0.0.11-py3.7.egg
Searching for flake8-docstrings==1.5.0
Best match: flake8-docstrings 1.5.0
Processing flake8_docstrings-1.5.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/flake8_docstrings-1.5.0-py3.7.egg
Searching for flake8-bugbear==19.8.0
Best match: flake8-bugbear 19.8.0
Processing flake8_bugbear-19.8.0-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/flake8_bugbear-19.8.0-py3.7.egg
Searching for emoji==0.5.4
Best match: emoji 0.5.4
Processing emoji-0.5.4-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/emoji-0.5.4-py3.7.egg
Searching for botocore==1.12.246
Best match: botocore 1.12.246
Processing botocore-1.12.246-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/botocore-1.12.246-py3.7.egg
Searching for boto3==1.9.246
Best match: boto3 1.9.246
Processing boto3-1.9.246-py3.7.egg

Using /Users/shaojie/src/ParlAI-1/.eggs/boto3-1.9.246-py3.7.egg
Searching for entrypoints<0.4.0,>=0.3.0
Best match: entrypoints 0.3
Processing entrypoints-0.3-py3.7.egg

Using /Users/shaojie/anaconda3/envs/newparlai/lib/python3.7/site-packages/entrypoints-0.3-py3.7.egg
running egg_info
writing parlai.egg-info/PKG-INFO
writing dependency_links to parlai.egg-info/dependency_links.txt
writing requirements to parlai.egg-info/requires.txt
writing top-level names to parlai.egg-info/top_level.txt
reading manifest file 'parlai.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'parlai.egg-info/SOURCES.txt'
running build_ext
test_verify_data (test_new_tasks.TestNewTasks) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.714s

OK
```

### CODE REVIEW

File: parlai/tasks/cornell_movie/agents.py [hunk 1]
```diff
@@ -63,20 +58,25 @@ def rebuild(entries):
...
+            return True
+
         # this shows conversations in both directions
         alternate = []
         for entry, new in super().setup_data(path):
             if new:
                 for i, e in enumerate(rebuild(alternate)):
-                    yield e, i == 0
+                    if is_valid(e):
+                        yield e, i == 0
```
Comment (stephenroller) [inline parlai/tasks/cornell_movie/agents.py]: Does this create a problem where the first entry is skipped and therefore you end up with two episodes running together?

File: parlai/tasks/cornell_movie/agents.py [hunk 2]
```diff
@@ -63,20 +58,25 @@ def rebuild(entries):
                         for i in range(len(entries) - 1)
                     ]
                 )
-                if len(entries[-1]) > 1 and entries[-1][1]:
-                    # add y_n => '', if last y avail
-                    new_list.append((entries[-1][1][0], None))
             return new_list
 
+        def is_valid(entry):
```
Comment (stephenroller) [inline parlai/tasks/cornell_movie/agents.py]: okay, since you're over here anyway, can you move rebuild and is_valid to be part of the class instead of closures?, i.e. `def _rebuild(self, entries)` and `def _is_valid(self, entry)`?

File: parlai/tasks/cornell_movie/agents.py (hunk 1 above)
Comment (stephenroller) [reply parlai/tasks/cornell_movie/agents.py]: Great, thanks for explaining. And the main teacher seems fine still?

File: parlai/tasks/cornell_movie/agents.py (hunk 2 above)
Comment (ShaojieJiang) [reply parlai/tasks/cornell_movie/agents.py]: Done

File: parlai/tasks/cornell_movie/agents.py (hunk 1 above)
Comment (ShaojieJiang) [reply parlai/tasks/cornell_movie/agents.py]: Not really. The DefaultTeacher has the same problems, except that empty source examples `['', y1]` only come from training data. I'll make a fix for it later.

Comment (stephenroller) [reply parlai/tasks/cornell_movie/agents.py]: Well one of our group has been using it lately and so I'm surprised they're not having a problem. You can submit a patch but it might have to wait.

Comment (ShaojieJiang) [reply parlai/tasks/cornell_movie/agents.py]: Sure. The problem of `None` target only appears occasionally, when a batch is full of such examples. I met that only on the test set, and I assume simple methods like using a different batch size can evade this problem. Anyways, we can leave this for future pull requests.

**Comment by stephenroller** [comment]
Do you mind running `python parlai/scripts/verify_data.py -t cornell_movies:double` with your patch for me? I think the test_new_data test didn't pick up the change.

**Comment by klshuster** [comment]
```
$ python parlai/scripts/verify_data.py -t cornell_movie:double
...
[ loaded 66460 episodes with a total of 133111 examples ]
1202s elapsed: {'exs': 176975, '%done': '132.95%', 'time_left': '-298s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 0, 'empty_string_label_candidates': 23, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}

$ python parlai/scripts/verify_data.py -t cornell_movie:double -dt valid
...
[ loaded 12806 episodes with a total of 22349 examples ]
24s elapsed: {'exs': 22349, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 0, 'empty_string_label_candidates': 1, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}

$ python parlai/scripts/verify_data.py -t cornell_movie:double -dt test
...

[ loaded 12790 episodes with a total of 22013 examples ]
24s elapsed: {'exs': 22013, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 0, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
```

not sure what happened here with going over 100% on train, but looks good to me

**Comment by klshuster** [comment]
thanks for making this change @ShaojieJiang!
